### PR TITLE
Moves linked overmap machines into a type list instead of discrete lists.

### PR DESCRIPTION
**Pull request by MistakeNot4892**
Also turns them into weakref lists.

### CODE REVIEW

File: code/modules/overmap/sectors.dm [hunk 1]
```diff
@@ -24,8 +24,27 @@
...
-	var/list/consoles
-	var/list/shield_generators
+	var/list/associated_machinery
+
+/obj/effect/overmap/visitable/proc/get_linked_machines_of_type(var/base_type)
+	return LAZYACCESS(associated_machinery, base_type)
+
+/obj/effect/overmap/visitable/proc/unregister_machine(var/obj/machinery/machine, var/base_type)
+	if(istype(machine))
+		base_type = base_type || (istype(machine) && (machine.base_type || machine.type))
```
Comment (comma) [inline code/modules/overmap/sectors.dm]: you /just/ typechecked it in line above

File: code/modules/overmap/sectors.dm [hunk 2]
```diff
@@ -24,8 +24,27 @@
...
+		if(islist(associated_machinery) && associated_machinery[base_type])
+			LAZYREMOVE(associated_machinery[base_type], machine)
+
+/obj/effect/overmap/visitable/proc/register_machine(var/obj/machinery/machine, var/base_type)
+	ASSERT(istype(machine))
+	if(!QDELETED(machine))
+		base_type = base_type || (istype(machine) && (machine.base_type || machine.type))
```
Comment (comma) [inline code/modules/overmap/sectors.dm]: here too, you crash if  it's not a machine

**Comment by comma** [comment]
Also doesn't actually use weakrefs anywhere, did you forget to push

```
[03:00:00] Runtime in sectors.dm,42: bad index
  proc name: register machine (/obj/effect/overmap/visitable/proc/register_machine)
  src: Tradeship Ocelot Alpha (/obj/effect/overmap/visitable/ship/tradeship)
  src.loc: 4-4 (4,4,7) (/turf/unsimulated/map)
  call stack:
  Tradeship Ocelot Alpha (/obj/effect/overmap/visitable/ship/tradeship): register machine(the advanced shield generator (/obj/machinery/power/shield_generator), /obj/machinery/power/shield_ge... (/obj/machinery/power/shield_generator))
  the advanced shield generator (/obj/machinery/power/shield_generator): update overmap shield list()
  the advanced shield generator (/obj/machinery/power/shield_generator): LateInitialize(1)
  Atoms (/datum/controller/subsystem/atoms): InitializeAtoms()
  Atoms (/datum/controller/subsystem/atoms): Initialize(818580)
  Atoms (/datum/controller/subsystem/atoms): DoInitialize(818580)
  Master (/datum/controller/master): Initialize(10, 0)
```

**Comment by MistakeNot4892** [comment]
Yeah forgot to push rip.

**Review by comma — APPROVED** [review]

**Pull request merged**